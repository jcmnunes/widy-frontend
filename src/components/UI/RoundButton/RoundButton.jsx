import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Spinner } from '..';

const StyledRoundButton = styled.button`
  display: flex;
  align-items: center;
  height: 36px;
  border-radius: 999px;
  border: none;
  background: none;
  cursor: pointer;
  font-size: 14px;
  color: ${props => props.theme.neutral700};
  padding: 0 8px;
  position: relative;
  opacity: ${props => (props.disabled && !props.loading ? 0.5 : 1)};

  span.children {
    margin: 0 4px;
  }

  &:hover {
    background: ${props => props.theme.neutral050};

    span.children {
      text-decoration: underline;
    }
  }

  &:focus {
    outline: none;
    box-shadow: 0 0 0 4px ${props => props.theme.blue100};
  }

  &:disabled {
    background: ${props => props.theme.neutral050};
    cursor: not-allowed;
  }

  .content {
    visibility: ${props => (props.loading ? 'hidden' : 'visible')};
    display: flex;
    align-items: center;
  }

  .spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    height: 18px;
  }
`;

const RoundButton = ({ loading, disabled, iconBefore, iconAfter, children, ...other }) => (
  <StyledRoundButton loading={loading} disabled={disabled || loading} {...other}>
    {loading && (
      <span className="spinner">
        <Spinner size="small" />
      </span>
    )}
    <div className="content">
      {iconBefore && iconBefore}
      {children && <span className="children">{children}</span>}
      {iconAfter && iconAfter}
    </div>
  </StyledRoundButton>
);

RoundButton.defaultProps = {
  loading: false,
  disabled: false,
  iconBefore: null,
  iconAfter: null,
  children: null,
};

RoundButton.propTypes = {
  loading: PropTypes.bool,
  disabled: PropTypes.bool,
  iconBefore: PropTypes.element,
  iconAfter: PropTypes.element,
  children: PropTypes.node,
};

export default RoundButton;
