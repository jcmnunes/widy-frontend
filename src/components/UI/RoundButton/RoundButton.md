# RoundButton

A special kind of button. It was designed originally to be used in the actions
bar inside the Board.

### Usage

```js
import { RoundButton } from 'components/UI';

<RoundButton iconBefore={<IconCheveronDown />} iconAfter={<IconUserCircle />}>
  Standup
</RoundButton>;
```

<!-- STORY -->

### Props

| Props        | propType | defaultValue | isRequired |
| ------------ | -------- | ------------ | ---------- |
| `loading`    | boolean  | `false`      |            |
| `disabled`   | boolean  | `false`      |            |
| `children`   | node     | `null`       |            |
| `iconBefore` | element  | `null`       |            |
| `iconAfter`  | element  | `null`       |            |

loading, disabled, iconBefore, iconAfter, children
