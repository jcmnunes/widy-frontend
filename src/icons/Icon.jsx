import React from 'react';
import PropTypes from 'prop-types';

const Icon = ({ icon, size, primaryColor, secondaryColor, className }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width={size}
    height={size}
    viewBox="0 0 24 24"
    className={className}
  >
    <path fill={primaryColor || 'currentColor'} className="primary" d={icon[0]} />
    <path fill={secondaryColor || 'currentColor'} className="secondary" d={icon[1]} />
  </svg>
);

Icon.defaultProps = {
  size: '24px',
  primaryColor: null,
  secondaryColor: null,
};

Icon.propTypes = {
  icon: PropTypes.arrayOf(PropTypes.string).isRequired,
  size: PropTypes.string,
  primaryColor: PropTypes.string,
  secondaryColor: PropTypes.string,
  className: PropTypes.string.isRequired,
};

Icon.CHART = [
  'M5 3h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5c0-1.1.9-2 2-2zm11 4a1 1 0 0 0-1 1v8a1 1 0 0 0 2 0V8a1 1 0 0 0-1-1z',
  'M8 11a1 1 0 0 1 1 1v4a1 1 0 0 1-2 0v-4a1 1 0 0 1 1-1zm4-2a1 1 0 0 1 1 1v6a1 1 0 0 1-2 0v-6a1 1 0 0 1 1-1',
];

Icon.LIGHT = [
  'M5 8a7 7 0 1 1 10.62 6l-.64 3.2a1 1 0 0 1-.98.8h-4a1 1 0 0 1-.98-.8L8.38 14A7 7 0 0 1 5 8zm12 0a5 5 0 0 0-5-5 1 1 0 0 0 0 2 3 3 0 0 1 3 3 1 1 0 0 0 2 0z',
  'M15 21a2 2 0 0 1-2 2h-2a2 2 0 0 1-2-2 1 1 0 0 1 0-2h6a1 1 0 0 1 0 2z',
];

Icon.PLAY = [
  'M12,2c5.5,0,10,4.5,10,10s-4.5,10-10,10S2,17.5,2,12S6.5,2,12,2z',
  'M15.51 11.14a1 1 0 0 1 0 1.72l-5 3A1 1 0 0 1 9 15V9a1 1 0 0 1 1.51-.86l5 3z',
];

Icon.STOP = [
  'M12,2c5.5,0,10,4.5,10,10s-4.5,10-10,10S2,17.5,2,12S6.5,2,12,2z',
  'M9.5 8.5h5a1 1 0 0 1 1 1v5a1 1 0 0 1-1 1h-5a1 1 0 0 1-1-1v-5a1 1 0 0 1 1-1z',
];

Icon.TIME = [
  'M12,2c5.5,0,10,4.5,10,10s-4.5,10-10,10S2,17.5,2,12S6.5,2,12,2z',
  'M13 11.59l3.2 3.2a1 1 0 0 1-1.4 1.42l-3.5-3.5A1 1 0 0 1 11 12V7a1 1 0 0 1 2 0v4.59z',
];

Icon.SURVEY = [
  'M5 5h14a2 2 0 0 1 2 2v13a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V7c0-1.1.9-2 2-2zm3 7a1 1 0 0 0 0 2h8a1 1 0 0 0 0-2H8zm0 4a1 1 0 0 0 0 2h4a1 1 0 0 0 0-2H8z',
  'M15 4a2 2 0 0 1 2 2v1a1 1 0 0 1-1 1H8a1 1 0 0 1-1-1V6c0-1.1.9-2 2-2 0-1.1.9-2 2-2h2a2 2 0 0 1 2 2z',
];

Icon.CHEVERON_DOWN = [
  'M15.3 10.3a1 1 0 0 1 1.4 1.4l-4 4a1 1 0 0 1-1.4 0l-4-4a1 1 0 0 1 1.4-1.4l3.3 3.29 3.3-3.3z',
  '',
];

Icon.CHEVERON_RIGHT = [
  'M10.3 8.7a1 1 0 0 1 1.4-1.4l4 4a1 1 0 0 1 0 1.4l-4 4a1 1 0 0 1-1.4-1.4l3.29-3.3-3.3-3.3z',
  '',
];

Icon.PRESENTATION = [
  'M11 18.62l-6.55 3.27a1 1 0 0 1-.9-1.78L11 16.38V5a1 1 0 0 1 2 0v11.38l7.45 3.73a1 1 0 0 1-.9 1.78L13 18.62V21a1 1 0 0 1-2 0v-2.38z',
  'M21 14a1 1 0 0 1 0 2H3a1 1 0 0 1 0-2V4a1 1 0 1 1 0-2h18a1 1 0 0 1 0 2v10z',
];

Icon.USER = [
  'M12 12a5 5 0 1 1 0-10 5 5 0 0 1 0 10z',
  'M21 20v-1a5 5 0 0 0-5-5H8a5 5 0 0 0-5 5v1c0 1.1.9 2 2 2h14a2 2 0 0 0 2-2z',
];

Icon.USER_CIRCLE = [
  'M3.66 17.52a10 10 0 1 1 16.68 0C19.48 16.02 17.86 16 16 16H8c-1.86 0-3.48.01-4.34 1.52z',
  'M3.66 17.52A5 5 0 0 1 8 15h8a5 5 0 0 1 4.34 2.52 10 10 0 0 1-16.68 0zM12 13a4 4 0 1 1 0-8 4 4 0 0 1 0 8z',
];

Icon.COG = [
  'M6.8 3.45c.87-.52 1.82-.92 2.83-1.17a2.5 2.5 0 0 0 4.74 0c1.01.25 1.96.65 2.82 1.17a2.5 2.5 0 0 0 3.36 3.36c.52.86.92 1.8 1.17 2.82a2.5 2.5 0 0 0 0 4.74c-.25 1.01-.65 1.96-1.17 2.82a2.5 2.5 0 0 0-3.36 3.36c-.86.52-1.8.92-2.82 1.17a2.5 2.5 0 0 0-4.74 0c-1.01-.25-1.96-.65-2.82-1.17a2.5 2.5 0 0 0-3.36-3.36 9.94 9.94 0 0 1-1.17-2.82 2.5 2.5 0 0 0 0-4.74c.25-1.01.65-1.96 1.17-2.82a2.5 2.5 0 0 0 3.36-3.36zM12 16a4 4 0 1 0 0-8 4 4 0 0 0 0 8z',
  'M12 10a2 2 0 1 1-2 2 2 2 0 0 1 2-2z',
];

export default Icon;
