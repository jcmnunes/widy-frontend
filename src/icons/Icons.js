/* eslint-disable max-len */
import React from 'react';
import PropTypes from 'prop-types';
import theme from '../styles/theme';

export const IconChart = ({ size, primaryColor, secondaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={primaryColor}
      d="M5 3h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5c0-1.1.9-2 2-2zm11 4a1 1 0 0 0-1 1v8a1 1 0 0 0 2 0V8a1 1 0 0 0-1-1z"
    />
    <path
      fill={secondaryColor}
      d="M8 11a1 1 0 0 1 1 1v4a1 1 0 0 1-2 0v-4a1 1 0 0 1 1-1zm4-2a1 1 0 0 1 1 1v6a1 1 0 0 1-2 0v-6a1 1 0 0 1 1-1z"
    />
  </svg>
);

IconChart.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
  secondaryColor: theme.neutral500,
};

IconChart.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
  secondaryColor: PropTypes.string,
};

export const IconLight = ({ size, primaryColor, secondaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={primaryColor}
      d="M5 8a7 7 0 1 1 10.62 6l-.64 3.2a1 1 0 0 1-.98.8h-4a1 1 0 0 1-.98-.8L8.38 14A7 7 0 0 1 5 8zm12 0a5 5 0 0 0-5-5 1 1 0 0 0 0 2 3 3 0 0 1 3 3 1 1 0 0 0 2 0z"
    />
    <path
      fill={secondaryColor}
      d="M15 21a2 2 0 0 1-2 2h-2a2 2 0 0 1-2-2 1 1 0 0 1 0-2h6a1 1 0 0 1 0 2z"
    />
  </svg>
);

IconLight.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
  secondaryColor: theme.neutral500,
};

IconLight.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
  secondaryColor: PropTypes.string,
};

export const IconTrophy = ({ size, primaryColor, secondaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={secondaryColor}
      d="M7 4v2H4v4c0 1.1.9 2 2 2h1v2H6a4 4 0 0 1-4-4V6c0-1.1.9-2 2-2h3zm10 2V4h3a2 2 0 0 1 2 2v4a4 4 0 0 1-4 4h-1v-2h1a2 2 0 0 0 2-2V6h-3zm-3 14h2a1 1 0 0 1 0 2H8a1 1 0 0 1 0-2h2a1 1 0 0 0 1-1v-3h2v3a1 1 0 0 0 1 1z"
    />
    <path fill={primaryColor} d="M8 2h8a2 2 0 0 1 2 2v7a6 6 0 1 1-12 0V4c0-1.1.9-2 2-2z" />
  </svg>
);

IconTrophy.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
  secondaryColor: theme.neutral500,
};

IconTrophy.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
  secondaryColor: PropTypes.string,
};

export const IconAdd = ({ size, primaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={primaryColor}
      fillRule="evenodd"
      d="M17 11a1 1 0 0 1 0 2h-4v4a1 1 0 0 1-2 0v-4H7a1 1 0 0 1 0-2h4V7a1 1 0 0 1 2 0v4h4z"
    />
  </svg>
);

IconAdd.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
};

IconAdd.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
};

export const IconCheveronRight = ({ size, primaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={primaryColor}
      d="M10.3 8.7a1 1 0 0 1 1.4-1.4l4 4a1 1 0 0 1 0 1.4l-4 4a1 1 0 0 1-1.4-1.4l3.29-3.3-3.3-3.3z"
    />
  </svg>
);

IconCheveronRight.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
};

IconCheveronRight.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
};

export const IconClose = ({ size, primaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={primaryColor}
      d="M15.78 14.36a1 1 0 0 1-1.42 1.42l-2.82-2.83-2.83 2.83a1 1 0 1 1-1.42-1.42l2.83-2.82L7.3 8.7a1 1 0 0 1 1.42-1.42l2.83 2.83 2.82-2.83a1 1 0 0 1 1.42 1.42l-2.83 2.83 2.83 2.82z"
    />
  </svg>
);

IconClose.defaultProps = {
  size: 24,
  primaryColor: theme.neutral400,
};

IconClose.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
};

export const IconUserCircle = ({ size, primaryColor, secondaryColor }) => (
  <svg xmlns="http://www.w3.org/2000/svg" width={size} height={size} viewBox="0 0 24 24">
    <path
      fill={secondaryColor}
      d="M3.66 17.52a10 10 0 1 1 16.68 0C19.48 16.02 17.86 16 16 16H8c-1.86 0-3.48.01-4.34 1.52z"
    />
    <path
      fill={primaryColor}
      d="M3.66 17.52A5 5 0 0 1 8 15h8a5 5 0 0 1 4.34 2.52 10 10 0 0 1-16.68 0zM12 13a4 4 0 1 1 0-8 4 4 0 0 1 0 8z"
    />
  </svg>
);

IconUserCircle.defaultProps = {
  size: 24,
  primaryColor: theme.neutral700,
  secondaryColor: theme.neutral300,
};

IconUserCircle.propTypes = {
  size: PropTypes.number,
  primaryColor: PropTypes.string,
  secondaryColor: PropTypes.string,
};
