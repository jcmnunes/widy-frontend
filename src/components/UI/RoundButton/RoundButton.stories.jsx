import React from 'react';
import { storiesOf } from '@storybook/react';
import { withDocs } from 'storybook-readme';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';
import { IconUserCircle, IconCheveronDown } from '../../../icons/Icons';

import { RoundButton, RoundButtonReadme } from '..';

storiesOf('UI components', module)
  .addDecorator(withKnobs)
  .addDecorator(withDocs(RoundButtonReadme))
  .add('RoundButton', () => {
    let label;
    let defaultValue;

    label = 'text';
    defaultValue = 'Standup';
    const buttonText = text(label, defaultValue);

    label = 'iconBefore';
    defaultValue = true;
    const iconBefore = boolean(label, defaultValue);

    label = 'iconAfter';
    defaultValue = true;
    const iconAfter = boolean(label, defaultValue);

    label = 'loading';
    defaultValue = false;
    const loading = boolean(label, defaultValue);

    label = 'disabled';
    defaultValue = false;
    const disabled = boolean(label, defaultValue);

    return (
      <RoundButton
        iconBefore={iconBefore ? <IconCheveronDown /> : null}
        iconAfter={iconAfter ? <IconUserCircle /> : null}
        loading={loading}
        disabled={disabled}
      >
        {buttonText}
      </RoundButton>
    );
  });
