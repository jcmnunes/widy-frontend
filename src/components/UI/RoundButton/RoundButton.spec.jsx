import React from 'react';
import { shallow } from 'enzyme';
import RoundButton from './RoundButton';
import { IconUserCircle, IconCheveronDown } from '../../../icons/Icons';

const TestRenderer = require('react-test-renderer');

describe('UI Components', () => {
  describe('RoundButton', () => {
    const props = {};

    it('should render as expected, without crashing (2)', () => {
      const tree = TestRenderer.create(<RoundButton {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should set the opacity to 0.5 when the button is disabled but not in a loading state', () => {
      const newProps = {
        ...props,
        loading: false,
        disabled: true,
      };
      const tree = TestRenderer.create(<RoundButton {...newProps} />).toJSON();
      expect(tree).toHaveStyleRule('opacity', '0.5');
    });

    it('should make content invisible when in a loading state', () => {
      const newProps = {
        ...props,
        loading: true,
      };
      const tree = TestRenderer.create(<RoundButton {...newProps} />).toJSON();
      expect(tree).toHaveStyleRule('visibility', 'hidden', { modifier: ' .content' });
    });

    it('should render nodes passed as children', () => {
      const children = 'Children text';
      const wrapper = shallow(<RoundButton {...props}>{children}</RoundButton>);
      expect(wrapper.find('.children')).toHaveLength(1);
      expect(wrapper.find('.children').text()).toBe('Children text');
    });

    it('should render the iconAfter and the iconBefore', () => {
      const children = 'Children text';
      const newProps = {
        ...props,
        iconBefore: <IconUserCircle />,
        iconAfter: <IconCheveronDown />,
      };
      const wrapper = shallow(<RoundButton {...newProps}>{children}</RoundButton>);
      expect(wrapper.find(IconUserCircle)).toHaveLength(1);
      expect(wrapper.find(IconCheveronDown)).toHaveLength(1);
    });

    it('should render a loading indicator if props.loading is true', () => {
      const newProps = {
        ...props,
        loading: true,
      };
      const wrapper = shallow(<RoundButton {...newProps} />);
      expect(wrapper.find('.spinner')).toHaveLength(1);
    });
  });
});
