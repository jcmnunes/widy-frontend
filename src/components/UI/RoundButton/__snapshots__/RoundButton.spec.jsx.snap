// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Components RoundButton should render as expected, without crashing (2) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 36px;
  border-radius: 999px;
  border: none;
  background: none;
  cursor: pointer;
  font-size: 14px;
  padding: 0 8px;
  position: relative;
  opacity: 1;
}

.c0 span.children {
  margin: 0 4px;
}

.c0:hover span.children {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 4px;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 .content {
  visibility: visible;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .spinner {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  height: 18px;
}

<button
  className="c0"
  disabled={false}
>
  <div
    className="content"
  />
</button>
`;
